### ISSUE TITLE
.type() is super slow

### ISSUE THREAD

**Issue by kytwb** [segmentio/nightmare#126]
As seen with `DEBUG=nightmare`:
```
  nightmare .goto() page loaded: success +6s
  nightmare .type() Lydia into input[name="firstname"] +503ms
  nightmare .type() Homenick into input[name="lastname"] +4ms
  nightmare .type() Hessel_Jeanne@Herzog.com into input[name="email"] +17s
  nightmare .type() 380 16 628 20 47 into input[name="phone"] +25s
  nightmare .type() 8140 Adrian Knolls Apt. 225 into input[name="address"] +25s
  nightmare .type() New Frances into input[name="city"] +24s
```

**Comment by reinpk** [comment]
that is super slow. do you have a nightmare script test case for a publicly available site?

**Comment by kytwb** [comment]
@reinpk Gist available here: https://gist.github.com/kytwb/de153003ef2c211774b9

**Comment by jbbarth** [comment]
Fwiw, I observed the same type of slowness on OSX with phantomjs *2.0* ; back to *1.9.x* and times were very reasonable again. Maybe the people involved in this issue could confirm/infirm that ?

**Comment by jbbarth** [comment]
More experiments:
- I experienced a very slow (> 20s) `type()` (and maybe `wait()`) on Linux as well with *any* version of Phantomjs (tested 1.9.0, 1.9.2, 1.9.7, 1.9.8, 2.0.1-devel)
- I confirm that reverting back to nightmare 1.6.1 made the performance be acceptable again, and starting from 1.6.2 it starts to be slow as hell (which may confirm @Quentin01's bisect)

Fwiw here's the problematic part of a debug output, don't know if it can help:
```
...
  nightmare .goto() page loaded: success +2s
  nightmare .type() user into input[name="identification"] +500ms
  nightmare .type() passwordz into input[name="password"] +7ms
  nightmare .click() on button[type="submit"] +21s
  nightmare .wait() for the next page load +25s
...
```

**Comment by kytwb** [comment]
Going to use @jbbarth fork while waiting for a fix.

**Comment by johntitus** [comment]
I believe this issue has the same root cause as this other one:
https://github.com/segmentio/nightmare/issues/123

It's the socket connection between the phantomjs module and it's dependency, shoe.

You can manually edit shoe yourself. Go into `node_modules/phantom/shoe/index.js` and change line 8 to read
```js
var server = sockjs.createServer({
        heartbeat_delay : 200
});
```

Can you try that out and see if it helps?

**Comment by kytwb** [comment]
@johntitus That fixed it for me. Problem is shoe is a dependence of phantom so would require to fork nightmare, fork phantom, fork shoe, fix, then cascade reference all the patched forks in each package.json etc.

**Comment by johntitus** [comment]
I've been looking at this module: https://github.com/baudehlo/node-phantom-simple

Could be a possible replacement for the `phantom` module. `node-phantom-simple` has no dependencies, except PhantomJS. It uses node's `child_process.spawn` instead of socket.io, so it should be a bit more solid.

However, the API is slightly different, so it wouldn't be a drop in replacement. Also, it seems to be intermittently maintained.

**Comment by kytwb** [comment]
I've been looking at this module: [johntitus/node-horseman](https://github.com/johntitus/node-horseman).

**Comment by sean-hill** [comment]
Any update to this? I'm running into the same issue. :+1:

**Comment by nickmerwin** [comment]
:+1:  can confirm @johntitus's solution worked.

**Comment by sean-hill** [comment]
@johntitus solution did work, but it stinks for us because I'd have to go in manually update all the node_modules on my production server...and then once they refresh with a `npm install` I'd have to do it again :frowning:

**Comment by nylen** [comment]
I was having the same problem.  Here's a way to :shit:-patch it from within your own app:

```js
// Monkey patch for https://github.com/segmentio/nightmare/issues/126
( function() {
	var sockjs = require( 'nightmare/node_modules/phantom/node_modules/shoe/node_modules/sockjs' );
	if ( ! sockjs._createServerOld ) {
		sockjs._createServerOld = sockjs.createServer;
		sockjs.createServer = function( options ) {
			if ( ! options ) {
				options = { };
			}
			if ( ! ( 'heartbeat_delay' in options ) ) {
				options.heartbeat_delay = 200;
			}
			return sockjs._createServerOld( options );
		};
	}
} )();
```

I'm not sure if this will still work if, for example, a compatible version of `shoe` is installed at a lower level of the modules tree.  Does anyone know?

**Comment by reinpk** [comment]
this is somewhat fixed in v2 #200 by switching from phantomjs to electron
